### DISCUSSION TITLE
Extremely simple configuration doesn't work (load order)

### DISCUSSION THREAD

**Discussion by grundic** [wbthomason/packer.nvim#777] [Q&A] [answered]
Hello,

I have started a migration to nvim+lua and decided to move from Plug to Packer as well.
My configuration is located here: https://github.com/grundic/vcsh-neovim-master

I have Packer cloned, the issue seems in the loading order. I would like to clone my configuration and bootstrap it with this command:

```bash
$ nvim --headless -c 'autocmd User PackerComplete quitall' -c 'PackerSync'
```

The problem I have is that the color scheme is not being installed, which leads to this error:

```bash
$ nvim --headless -c 'autocmd User PackerComplete quitall' -c 'PackerSync'
Error detected while processing /usr/local/home/cgi/.config/nvim/init.lua:
E5113: Error while calling lua chunk: /usr/local/home/cgi/.config/nvim/lua/settings.lua:48: Vim(colorscheme):E185: Cannot find color scheme 'gruvbox'
stack traceback:
        [C]: in function 'cmd'
        /usr/local/home/cgi/.config/nvim/lua/settings.lua:48: in main chunk
        [C]: in function 'require'
        /usr/local/home/cgi/.config/nvim/init.lua:18: in main chunkError detected while processing command line:
E492: Not an editor command: PackerSync
```

Here is my ~/.config/nvim/init.lua (simplified):

```lua
require('settings')
require('autocommands')
require('plugins')
```

The `setting.lua` contains colorsheme selection: `cmd [[colorscheme gruvbox]]` and the `plugins.lua` lists a dependency on that colorscheme: `use 'morhetz/gruvbox'`.

I tried changing the order of `require('settings')` and `require('plugins')`, making sure that plugins are first, but it still doesn't work. I haven't found anything related in the documentation.

There were several question, which looks similar, but I don't understand how should I proceed:

  * https://github.com/wbthomason/packer.nvim/discussions/651
  * https://github.com/wbthomason/packer.nvim/discussions/565
  * https://www.reddit.com/r/neovim/comments/mohogr/neovim_lua_config_cant_install_plugins_when/
  
Can someone please helm me to fix this problem?
Thank you.

**Comment by wbthomason** [comment]
I believe that this is a problem only with bootstrapping - the colorscheme loading error seems to prevent the rest of `init` from running (based on the `Not an editor command: PackerSync` you also get), which prevents the plugin from being installed. I am a bit surprised that reordering didn't work, but this is probably due to Neovim only checking for colorschemes at start. When you reorder to have `plugins` first, then run once, then open Neovim again, does the colorscheme load as expected?

As @mjlbach mentioned on Matrix, you can use a trick wherein you set an environment variable to signal that you're in an installation run and skip the parts of the config that rely on yet-uninstalled dependencies. If you only ever use headless Neovim for installs/bootstrapping, a slightly easier version of this trick checks the number of attached UIs (`vim.api.nvim_list_uis()`).

Another possible workaround is to wrap the `cmd` call with `pcall`, which should prevent the error from breaking the later use of `PackerSync`, and get the plugins to install. I haven't tested this myself, though.

**Comment by grundic** [comment, marked as answer]
I've come with simple WTR solution:

```bash
$ cat <<EOF > simple.lua
print("Test configuration has sucessfully been loaded")
require('packer').startup(function()
  use 'wbthomason/packer.nvim'
  use 'morhetz/gruvbox'
end)

cmd [[colorscheme gruvbox]]
EOF

# make sure the gruvbox is not installed, remove generated cache
$ rm -rf ~/.local/share/nvim/site/pack/packer/start/gruvbox 
$ rm -f ~/.config/nvim/plugin/packer_compiled.lua

# try to start neovim
$ nvim -u simple.lua

Test configuration has sucessfully been loaded
Error detected while processing /usr/local/home/cgi/simple.lua:
E5113: Error while calling lua chunk: simple.lua:11: attempt to call global 'cmd' (a nil value)
stack traceback:
        simple.lua:11: in main chunk
Press ENTER or type command to continue
```

I tried using `pcall(cmd, [[colorscheme gruvbox]])`, which works fine. But that means that users have to wrap other potential places with `pcall` to make sure that it won't fail during initial setup (e.g. all plugins settings have to be wrapped with it). 
It might be a limitation of Lua itself, which is unfortunate.

Actually, after wrinthing this I came with an idea:

```bash
$  nvim -u .config/nvim/lua/plugins.lua --headless -c 'autocmd User PackerComplete quitall' -c 'PackerSync'
```

Aaand it worked like a charm 🥳  If all plugin requirements are collected in a single file, we can load it for bootstrap, ignoring all other settings. 
@wbthomason, do you think it is a good solution for a given problem?

**Reply by wbthomason** [reply]
Ah, that's quite clever! I like that - let's add it to the docs.

Also, for future reference: I believe the reason you get the `nil global` error is because, until you bind it, `cmd` is not in the global namespace - it's `vim.cmd` instead.

For most plugin settings, you can use `packer`'s `config` key feature to specify code that will only run after the plugin is installed and loaded. I didn't recommend this for the colorscheme because some users have reported issues with setting their colorscheme later in their startup (i.e. when `packer_compiled` is sourced).

**Reply by grundic** [reply]
> Also, for future reference: I believe the reason you get the nil global error is because, until you bind it, cmd is not in the global namespace - it's vim.cmd instead.

Ah, right. In the original script I made an alias `local cmd = vim.cmd` and forgot about it in the example. Yes, after calling it like `vim.cmd` I got a proper error 

```
E5113: Error while calling lua chunk: simple.lua:11: Vim(colorscheme):E185: Cannot find color scheme 'gruvbox'
```

> let's add it to the docs.

Sweet, will prepare a PR 🤓
